---
output: 
  github_document:
    html_preview: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

fqdecontamr filters fastq files to remove reads identified as potential
contaminants by [decontam](https://benjjneb.github.io/decontam/).

## Installation

You can install the development version of fqdecontamr from
[GitHub](https://github.com/wilkox/fqdecontamr) with:

``` r
devtools::install_github("wilkox/fqdecontamr")
```

## How to use

fqdecontamr requires two input files:

- The decontam output, in the form of a Microsoft Excel spreadsheet (`*.xlsx`)
- The fastq file you want to filter (gzip compressed files are accepted)

The main function provided by fqdecontamr is called `decontaminate()`. It takes
as arguments the paths to the two input files listed above ( `decontam_file`
and `fastq_file`), a third argument `index_dir` specifying a directory in which
to cache downloaded genomes and bowtie2 index files, and a fourth argument
`out_file` specifying where to write the filtered fastq file. `decontaminate()`
returns  a [tibble](https://tibble.tidyverse.org) listing the reads that were
removed and the contaminant species to which they aligned.
